server/src/main/java/org/elasticsearch/action/admin/indices/rollover/LazyRolloverAction.java

```diff
@@ -124,7 +145,61 @@ protected void masterOperation(
             newRolloverRequest.setIndicesOptions(rolloverRequest.indicesOptions());
             RolloverTask rolloverTask = new RolloverTask(newRolloverRequest, null, trialRolloverResponse, null, listener);
-            submitRolloverTask(rolloverRequest, source, rolloverTask);
+            lazyRolloverTaskQueue.submitTask(source, rolloverTask, rolloverRequest.masterNodeTimeout());
+        }
+    }
+
+    /**
+     * Extends the {@link TransportRolloverAction.RolloverExecutor} and delegates all the execution to it
+     * after it confirms that the rollover should be executed. A lazy rollover should be executed
+     * iff the rolloverOnWrite flag is true on the requested target. Otherwise, we assume that a different
+     * event has triggered the rollover and this one is not necessary anymore.
+     */
+    static class LazyRolloverExecutor extends TransportRolloverAction.RolloverExecutor {
+        LazyRolloverExecutor(
+            ClusterService clusterService,
+            AllocationService allocationService,
+            MetadataRolloverService rolloverService,
+            ThreadPool threadPool
+        ) {
+            super(clusterService, allocationService, rolloverService, threadPool);
+        }
+
+        @Override
+        public ClusterState executeTask(
+            ClusterState currentState,
+            List<MetadataRolloverService.RolloverResult> results,
+            TaskContext<TransportRolloverAction.RolloverTask> rolloverTaskContext,
+            AllocationActionMultiListener<RolloverResponse> allocationActionMultiListener
+        ) throws Exception {
+            final var rolloverTask = rolloverTaskContext.getTask();
+            final var rolloverRequest = rolloverTask.rolloverRequest();
+
+            final IndexAbstraction rolloverTargetAbstraction = currentState.metadata()
+                .getIndicesLookup()
+                .get(rolloverRequest.getRolloverTarget());
+            assert rolloverTargetAbstraction.getType() == IndexAbstraction.Type.DATA_STREAM : "Lazy rollover only applied to data streams";
+            final DataStream dataStream = (DataStream) rolloverTargetAbstraction;
+            final DataStream.DataStreamIndices targetIndices = dataStream.getDataStreamIndices(rolloverRequest.targetsFailureStore());
+            if (targetIndices.isRolloverOnWrite() == false) {
+                rolloverTaskContext.success(() -> rolloverTask.listener().onResponse(noopLazyRolloverResponse(targetIndices)));
+                return currentState;
+            }
+            return super.executeTask(currentState, results, rolloverTaskContext, allocationActionMultiListener);
         }
     }
+
+    private static RolloverResponse noopLazyRolloverResponse(DataStream.DataStreamIndices indices) {
+        String latestWriteIndex = indices.getWriteIndex().getName();
+        return new RolloverResponse(
+            latestWriteIndex,
+            latestWriteIndex,
+            Map.of("rollover_on_write", false),
```

Comment:
I am not set on adding this, but I thought it might be nice for debugging purposes.